crates/runtime/src/trampolines/aarch64.rs

```diff
@@ -0,0 +1,155 @@
+use crate::backtrace::{Backtrace, Frame};
+use std::ops::ControlFlow;
+
+pub struct TrapState {
+    pc: usize,
+    sp: usize,
+    fp: usize,
+    lr: usize,
+}
+
+impl TrapState {
+    #[cfg(unix)]
+    pub unsafe fn from_ucontext(cx: *mut libc::c_void, _signum: libc::c_int) -> TrapState {
+        let cx = &*(cx as *const libc::ucontext_t);
+        TrapState {
+            pc: cx.uc_mcontext.pc as usize,
+            sp: cx.uc_mcontext.sp as usize,
+            fp: cx.uc_mcontext.regs[29] as usize,
+            lr: cx.uc_mcontext.regs[30] as usize,
+        }
+    }
+
+    pub fn pc(&self) -> usize {
+        self.pc
+    }
+
+    pub fn fp(&self) -> usize {
+        self.fp
+    }
+
+    pub unsafe fn trace_initial(
+        &self,
+        first_wasm_sp: usize,
+        mut f: impl FnMut(Frame) -> ControlFlow<()>,
+    ) -> ControlFlow<(), Option<(usize, usize)>> {
+        // TODO: needs a lot of comments
+        f(Frame {
+            pc: self.pc,
+            fp: Some(self.fp),
+        })?;
+        f(Frame {
+            pc: self.lr,
+            fp: None,
+        })?;
+        if first_wasm_sp == self.sp {
+            return ControlFlow::Continue(None);
+        }
+        ControlFlow::Continue(Backtrace::trace_one_wasm_frame(self.fp, first_wasm_sp))
+    }
+}
+
+std::arch::global_asm!(
+    r#"
+.globl host_to_wasm_trampoline
+host_to_wasm_trampoline:
+    // Load the pointer to `VMRuntimeLimits` in `x9`.
+    ldur x9, [x1, #8]
+
+    // Check to see if callee is a core `VMContext` (MAGIC == "core").
+    ldur w10, [x0]
+    mov  w11, #0x6f63
+    movk w11, #0x6572, lsl #16
+    cmp  w10, w11
+
+    // Store the last Wasm SP into the `last_wasm_entry_sp` in the limits, if
+    // this was core Wasm, otherwise store an invalid sentinal value.
+    mov  x12, #-1
+    mov  x13, sp
+    csel x12, x13, x12, eq // TODO: is `eq` right?
+    stur x12, [x9, #40]
+
+    // Tail call to the callee function pointer in the vmctx.
+    ldur x9, [x1, #16]
+    br   x9
+"#
```

Comment:
Similarly to line 56, end with `.cfi_endproc`.

Reply:
All the CFI stuff will be fixed with using the macros in the `wasmtime-fibers` crate.